docs/scaling-hive.md

```diff
@@ -0,0 +1,47 @@
+
+# Horizontal vs. Vertical Scale
+
+With the exception of install pods (used only when clusters are installing), Hive 1.x is not horizontally scalable at the worker level. Most of the work Hive does happens in the hive-controllers pod, which is one single pod on one single worker. This means that when no installs are running, if you have a cluster with 10 workers, 9 of the workers are very bored. Hive clusters are prime candidates for using worker autoscaling. Keep the worker count as low as you can, but allow bursts of concurrent installs to call for temporary workers to spin up.
+
+In AWS, Hive performs best on c instance types. M instances are fine, but c instances are better.
+
+In general, the faster your etcd storage, the faster Hive will perform. c5.4xlarge has twice the EBS bandwidth as c5.2xlarge, and so Hive performs a bit better on c5.4xlarge (although in our opinion, not so much faster that c5.4xlarge is worth the premium over c5.2xlarge). In AWS, we recommend using io1 with 1200 iops for your masters.
+
+If you need to scale out the number of clusters you need to manage, use multiple Hive clusters -- approximately 1 Hive per 1000 clusters managed. Or fewer if you prefer to keep the blast radius smaller.
+
+## Install Pods
+
+Hive 1.x request 800 Mib of memory for each install pod. If you use m5.xlarge workers, you can support about (15 Gib / 800 Mib) install pods per worker -- so about 16. If you need to support more concurrent installs, you can use more workers, and/or workers with more memory. Install pods use barely any CPU.
+
+## Blocking I/O
+
+hive-controllers (where the syncset controllers run) uses blocking i/o. We may move to scale-out or non-blocking i/o in the future, but for now you should be aware that communication (mostly syncsets) to managed clusters can block other things.
+
+## Threads
+
+Hive supports configuring the number of goroutines per controller by editing values in HiveConfig. See [Using Hive](using-hive.md) for documentation on this. In general, set the number of goroutines to roughly the number of vCPUs on your workers. For c5.2xlarge, we use 10 goroutines. If you run Hive on larger hardware, you need to bump up the goroutines. If you have only 10 goroutines but 100 vCPUs, 90 of the vCPUs will be mostly unused. If Hive manages clusters that are on slow networks or have connectivity issues, you may want to use more goroutines to work around Hive's use of blocking i/o. Don't use too many goroutines, or that will actually cause a slowdown (i.e. don't just set it to 1000 or something).
```

Comment:
Are you recommending setting numvcpus for *each* controller, meaning the total goroutines is numvcpus*numcontrollers? This paragraph needs to be clarified in regards to this in a few places.

Do we still default to 5? If so that should be documented. Are we actively running 10 on c5.2xlarge anywhere? Also we should note that's AWS c5.2xlarge.

I'm really unsure about linking to vcpus so heavily here especially given we just said we're IO blocking. This feels somewhat inaccurate, but maybe you're implying it's just a good general guideline?

Reply:
In a busy hive cluster, over the life of the cluster, the ss controller is the thing that does most of the work. The other controllers are bored compared to ss. So I'm simplifying things and only thinking about ss.

The goal of anyone looking to push hive should be to the keep CPU utilization for the worker as close to 100% as possible.

I found that we get the highest CPU utilization (and thus best overall performance) when we roughly match the number of ss threads to the number of vCPUs on the worker. If we use less, the CPU is bored. If we have 36 CPUs but only 20 threads, CPU utilization is usually 50%. If we match (40 threads for 36 vCPU), CPU utilization is high 90's. If we use too many threads (100 threads for 36 vCPU), we still get high CPU utilization, but the process actually slows down. I assume this is golang overhead of too many threads / thread management.

Generally, blocking i/o is not a problem. It should be rare for hive to manage a lot of clusters that are on slow networks, unreachable, etc. So I simplify things and don't really take that into consideration in this section. If you're in a special scenario where you *do* manage a lot of clusters that are frequently unreachable, you would frequently have several threads blocked on i/o hangs -- doing nothing. CPU utilization would drop. So you could compensate by adding in a couple more threads.

Does this make sense?

Reply:
Makes sense but it needs to be clarified here that we're talking about clustersync controller specifically. I'm wouldn't be crazy about this advice for all controllers.

Reply:
I took large portions of my above comment and added it. Hopefully it is clearer now.